cpp/serve/engine_actions/batch_draft.cc

```diff
@@ -122,7 +145,18 @@ class BatchDraftActionObj : public EngineActionObj {
 
         // - Add draft token to the state.
         for (int i = 0; i < num_rsentries; ++i) {
-          mstates[i]->AddDraftToken(sample_results[i], prob_dist[i]);
+          // - Slice hidden_states_for_sample
+          NDArray last_hidden_on_device = NDArray::Empty(
+              {hidden_states_nd->shape[2]}, hidden_states_nd->dtype, hidden_states_nd->device);
+          int64_t ndata = hidden_states_nd->shape[2];
+          const int16_t* __restrict p_hidden =
+              static_cast<int16_t*>(__builtin_assume_aligned(hidden_states_nd->data, 2)) +
+              (i * ndata);
+
+          last_hidden_on_device.CopyFromBytes(p_hidden, ndata * sizeof(int16_t));
```

Comment:
What does int16 mean here? Isn't the hidden state floating point?

Reply:
I think logits are fp32, while hidden states are fp16. When I use float type, an error occurs here.

Reply:
This is also assumption we should avoid. In this case let's use `char` or `unsigned char` and copy raw bytes.